site/_posts/2019-10-09-Custom-E2E-Image.md

```diff
@@ -0,0 +1,105 @@
+
+It’s most significant feature is the new environment variables, which can be used to set arbitrary flags when  the tests are invoked. For example, to set the GCE regions mentioned above, you run Sonobuoy with the following command:
+
+```
+$ sonobuoy run --plugin-env=e2e.E2E_USE_GO_RUNNER=true \
+--plugin-env=e2e.E2E_PROVIDER=gce \
+--plugin-env=e2e.E2E_EXTRA_ARGS=”--gce-zone=foo --gce-region=bar”
+```
+
+By default, the extra arguments are split by spaces, so if you have values that require spaces, you can also set a [custom separator](https://github.com/kubernetes/kubernetes/blob/master/cluster/images/conformance/go-runner/cmd_test.go#L101-L110) with `E2E_EXTRA_ARGS_SEP`.
+
+The only downside to this approach is that it is only available for clusters using Kubernetes 1.16.0. If you have an older cluster and want to utilize this feature, you will have to build your own test image (otherwise, the image will run the wrong set of tests for your cluster).
+
+# Building Your Own Test Image
+
+I’ve often referred people to the Kubernetes repo (and its [instructions](https://github.com/kubernetes/kubernetes/tree/master/cluster/images/conformance#how-to-release-by-hand)) when they need help building a custom test image.  That approach would surely work, but it involves actually building the Kubernetes tests yourself, which may tax your local machine or be difficult to set up the first time.
+
+For most users, I’d recommend taking a much simpler approach: Just built a new image based on the existing one and swap out the script as needed.
+
+By building the existing test image, the Kubernetes build process has already done the hard work: Built the tests, downloaded ginkgo, loaded everything into the right places, and included the script to execute the tests. All you have to do is tweak a single script.
+
+So, for instance, if you want to support the provider-specific values, all you have to do is:
+Create your custom script (usually just a slight modification of the existing one)
+Create a new image and make it available (a two-line Dockerfile)
+Create a custom plug-in definition that will use your image instead of the upstream Kubernetes conformance image (a single command)
+Run Sonobuoy with the custom plug-in
+
+Almost all of this can be entirely scripted:
+
+```
+# Be sure to push images to the correct registry and use the right version of tests.
+$ export REGISTRY=schnake
+$ export k8sVersion=$(kubectl version -o json|jq .serverVersion.gitVersion -r) 
+
+# Download the existing run script as a starting place and change its mode to be executable.
+$ curl https://raw.githubusercontent.com/kubernetes/kubernetes/master/cluster/images/conformance/run_e2e.sh -o run.sh && chmod +x run.sh
```

Comment:
Should the downloaded script come from the matching Kubernetes version instead?

Reply:
The script doesn't change much since it moved upstream; typically just bug fixes and such. I don't think it is a big deal to take the one from master.

Reply:
That's what I guessed but just wanted to check :)